src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleShardSyncLeaderDecider.java

```diff
@@ -0,0 +1,184 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+import com.amazonaws.services.kinesis.leases.exceptions.DependencyException;
+import com.amazonaws.services.kinesis.leases.exceptions.InvalidStateException;
+import com.amazonaws.services.kinesis.leases.exceptions.ProvisionedThroughputException;
+import com.amazonaws.services.kinesis.leases.impl.KinesisClientLease;
+import com.amazonaws.services.kinesis.leases.interfaces.ILeaseManager;
+import com.amazonaws.util.CollectionUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * An implementation of the {@code LeaderDecider} to elect leader(s) based on workerId.
+ * Leases are shuffled using a predetermined constant seed so that lease ordering is
+ * preserved across workers.
+ * This reduces the probability of choosing the leader workers co-located on the same
+ * host in case workerId starts with a common string (e.g. IP Address).
+ * Hence if a host has 3 workers, IPADDRESS_Worker1, IPADDRESS_Worker2, and IPADDRESS_Worker3,
+ * we don't end up choosing all 3 for shard sync as a result of natural ordering of Strings.
+ * This ensures redundancy for shard-sync during host failures.
+ */
+public class DeterministicShuffleShardSyncLeaderDecider implements LeaderDecider {
+    // Fixed seed so that the shuffle order is preserved across workers
+    static final int DETERMINISTIC_SHUFFLE_SEED = 1947;
+    static final int PERIODIC_SHARD_SYNC_MAX_WORKERS_DEFAULT = 1; //Default for KCL.
+    private static final Log LOG = LogFactory.getLog(DeterministicShuffleShardSyncLeaderDecider.class);
+    private static final int LEADER_DECIDER_THREAD_COUNT = 1;
+    private static final long ELECTION_INITIAL_DELAY_MILLIS = 60 * 1000;
+    private static final long ELECTION_SCHEDULING_INTERVAL = 5 * 60 * 1000;
+    private static final int AWAIT_TERMINATION_MILLIS = 5000;
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    private final KinesisClientLibConfiguration config;
+    private final ILeaseManager<KinesisClientLease> leaseManager;
+    private final int numPeriodicShardSyncWorkers;
+    private final ScheduledExecutorService leaderElectionThreadPool;
+
+    private volatile Set<String> leaders;
+
+    /**
+     * Create an instance using the default periodic shard sync worker count (1).
+     * This constructor is package-private to ensure external users use the public
+     * constructor and are thereby aware of the number of periodic shard sync workers
+     * they want in their application.
+     *
+     * @param config       KinesisClientLibConfiguration instance
+     * @param leaseManager LeaseManager instance.
+     */
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
```

Comment:
Is there a reason for new class to already have chained constructors?

Reply:
Why not?

Reply:
Are the chained constructors used for a reason? Just because we can add code doesn't mean we do. Its for maintenance purposes, if its a new class we can create the exact constructors that we need.

Reply:
We have 3 constructors - a default one with all default params (except config and lease manager), an overriding constructor that allows setting the number of shard sync workers (in case you want to go over 1, or allow customers override it, in future), and an all-args constructor for unit tests. I chained them to ensure less repetitive code. If you think any of these are un-necessary, or would like to structure it differently, I can make that change.

Reply:
Changed it to have just one constructor.